tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+}
+
+// Eval implements Expression's Eval()
+func (u *unOp) Eval(target *prog.Target) uint64 {
+	opEval := u.operand.Eval(target)
+	switch u.op {
+	case onescompOp:
+		return ^opEval
+	case negOp:
+		return -opEval
+	default:
+		log.Fatalf("Unsupported Unop Op: %d", u.op)
+	}
+	return 0
+}
+
+// String implements IrType String()
+func (u *unOp) String() string {
+	return fmt.Sprintf("op1: %v operand: %v\n", u.operand, u.op)
+}
+
+// DynamicType represents cases where strace shows the before system call and after
+// E.g. <... getsockname resumed> {sa_family=AF_INET6,..., sin6_scope_id=0}, [112->28]
+type DynamicType struct {
+	BeforeCall Expression
+	AfterCall  Expression
+}
+
+func newDynamicType(before, after IrType) *DynamicType {
+	return &DynamicType{BeforeCall: before.(Expression), AfterCall: after.(Expression)}
+}
+
+// String implements IrType String()
+func (d *DynamicType) String() string {
+	return d.BeforeCall.String()
+}
+
+// Name implements IrType Name()
+func (d *DynamicType) Name() string {
+	return "Dynamic Type"
+}
+
+// Garbage type used to parse content inside of ending parenthesis from strace
+// E.g. access(....) = -1 ENOENT (No such file or directory). The stuff in the ending parenthesis
+// will get parsed into this type
+type parenthetical struct {
+	tmp string
+}
+
+func newParenthetical() *parenthetical {
+	return &parenthetical{tmp: "tmp"}
+}
+
+// BufferType contains strings
+type BufferType struct {
+	Val string
+}
+
+func newBufferType(val string) *BufferType {
+	return &BufferType{Val: val}
+}
+
+// Name implements IrType Name()
+func (b *BufferType) Name() string {
+	return bufferTypeName
+}
+
+// String implements IrType String()
+func (b *BufferType) String() string {
+	return fmt.Sprintf("Buffer: %s with length: %d\n", b.Val, len(b.Val))
+}
+
+type intType struct {
+	Val int64
+}
+
+func newIntType(val int64) (typ *intType) {
+	return &intType{Val: val}
+}
+func (i *intType) eval(target *prog.Target) uint64 {
+	return uint64(i.Val)
+}
+
+func (i *intType) string() string {
+	return strconv.Itoa(int(i.Val))
+}
+
+// Flags contains set of flagTypes. Most of the time will contain just 1 element
+type Flags []*flagType
+
+// Ints Contains set of intTypes. Most of the time will contain just 1 element
+type Ints []*intType
+
+// NewIntsType - Constructor
+func NewIntsType(vals []int64) Ints {
+	var ints []*intType
+	for _, v := range vals {
+		ints = append(ints, newIntType(v))
+	}
+	return ints
+}
+
+// Eval implements Expression's Eval()
+func (f Flags) Eval(target *prog.Target) uint64 {
+	a := make([]*flagType, len(f))
+	copy(a, f)
+	val := uint64(0)
+	for len(a) > 0 {
+		val |= a[0].eval(target)
+		a = a[1:]
+	}
+	return val
+}
+
+// Name implements IrType Name()
+func (f Flags) Name() string {
+	return exprTypeName
+}
+
+// String implements IrType String()
+func (f Flags) String() string {
+	a := make([]*flagType, len(f))
+	var buf bytes.Buffer
+	copy(a, f)
```

Comment:
Copying is unnecessary and this looks overly complex. This would do:
```
	s := ""
	for _, v := range f {
		s += " " + f.string()
	}
	return s[1:]
```

If flags can be empty (can it?), then we could add the following in the beginning:
```
	if len(f) == 0 {
		return ""
	}
```

Reply:
No flags should not be empty ever.